mewarpx/mewarpx/utils_store/profileparser.py

```diff
@@ -0,0 +1,148 @@
+            "frame" : {"name" : "exclusive_function_profiles"},
+            "metrics" : {},
+            "children" : self.exclusive_profiles
+        })
+
+        self.profile_dicts.append({
+            "frame" : {"name" : "inclusive_function_profiles"},
+            "metrics" : {},
+            "children" : self.inclusive_profiles
+        })
+        total_count = excl_count + incl_count
+
+        print(f"Finished parsing profile output! Parsed {total_count} profiling strings.")
+
+    def parse_section_of_profiling_output(self, section):
+        '''
+        Parses either the exclusive or inclusive sections of the TinyProfiler output
+
+        Parameters
+        ----------
+
+        section (string): The section to parse the profiling output,
+            either 'excl' for exclusive or 'incl' for inclusive.
+
+        Returns
+        -------
+
+        The number of lines that were parsed
+        '''
+        if section != "excl" and section != "incl":
+            raise RuntimeError(f"'section' was not either 'excl' or 'incl', instead was {section}")
+
+        count = 0
+
+        for line in self.lines:
+            if "------" in line:
+                break
+            # these lines are of the form "Function()  ncalls  min  avg  max  max_percent"
+            # with more spaces than are here
+
+            # replace the multiple spaces with a single space
+            line = re.sub("\s+", " ", line)
+            values = line.split(" ")
+
+            # work backwards through the values in order to capture function names with spaces.
+            max_percent = float(values[-1].replace("%", ""))
+            max_time = float(values[-2])
+            avg = float(values[-3])
+            min_time = float(values[-4])
+            n_calls = int(values[-5])
+            name = "_".join(values[0:-5])
+
+            dict = {
+                "frame" : {"name" : name},
+                "metrics" : {
+                    "n_calls" : n_calls,
+                    f"{section}_min" : min_time,
+                    f"{section}_avg" : avg,
+                    f"{section}_max" : max_time,
+                    "max_percent" : max_percent
+                }
+            }
+
+            if section == "excl":
+                self.exclusive_profiles.append(dict)
+            elif section == "incl":
+                self.inclusive_profiles.append(dict)
+
+            count = count + 1
+
+        return count
+
+    def write_file(self):
+        if not self.profile_dicts:
+            raise RuntimeError("write_file called before profile data was parsed!")
+
+        print("Writing profile data json file...")
+        with open("profile_data.json", "w") as data_file:
+            data_file.write(json.dumps(self.profile_dicts, indent=4))
+
+        print("Finished writing json file as 'profile_data.json'.")
+
+def main():
+    if len(sys.argv) != 2:
+        raise RuntimeError(f"Usage: {sys.argv[0]} <path to stdout file>")
```

Comment:
Can you add the ability to pass the stdout file to the parser through the main() function call as well as through the command line? That way, we can just call this main function at the end of a warpx script, so it can automatically generate the json, instead of having to do it by hand. One way I can think of doing it is to add a path argument to `main()` remove the sys.argv check from `main`. Instead add the check for command line arguments to the `__name__ == "__main__"` part, and pass `sys.argv[1]` to the main function as an argument if it exists.

Reply:
This is a really really great idea, definitely makes using the script easy if I can get it to work

Reply:
@KZhu-ME The problem with this is that even the very end of a python script is before the TinyProfiler output is printed. It would be possible to call that from _libwarpx, but _libwarpx doesn't know about mewarpx stuff.

Reply:
Ooooh yeah I see, that wouldn't work. I still think it's worth moving the sys.argv check to  the `if name == main:` block, since it seems cleaner that way.

Reply:
If you really wanted to do that you could add this to the bottom of mwxrun.py (you need to import atexit as well)

```
def exit_handler():
    atexit.unregister(_libwarpx.finalize)
    _libwarpx.finalize()
    print('My application is ending!')
    print('Insert profiling summary here.')

atexit.register(exit_handler)
```

Reply:
Wow I've never heard of this. Is this something you'd be interested in? Automatically writing the JSON file seems like it would be nice to me.

Reply:
Yeah it sounds like a worthwhile change to me as well.